Added support for error handling

dibyadas opened the pull request:
> This PR deals with issue #16 

Custom exceptions for `python_http_client` have been added. Only few exceptions have been added. 
Others will be added soon.

SendGridDX commented:
[![CLA assistant check](https://cla.sendgrid.com/pull/badge/not_signed)](https://cla.sendgrid.com/sendgrid/python-http-client?pullRequest=17) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.sendgrid.com/sendgrid/python-http-client?pullRequest=17) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.sendgrid.com/check/sendgrid/python-http-client?pullRequest=17) it.</sub>

dibyadas commented:
Hi @thinkingserious !  I have added custom exception handling. Do suggest changes for improvements.
(P.S - Sorry for the delay. I got caught up in some work)

dibyadas commented:
Hi @w- ! Have a look at this. What do you think?  :)

w- commented:
Hey guys, i'll be out of town on vacation for the next few days. won't be able to respond on this topic till then.

thinkingserious commented:
Thanks for the heads up @w- and I hope you enjoy the vacation :)

thinkingserious commented:
@w-,

Do you have time to take a look? Thanks!

dibyadas commented:
@thinkingserious,

In the meantime, how about I add tests for this? 
Where do you think would be a good place to add?
Should I add it in the existing one or a different file for testing exceptions?

With Regards,
Dibya Prakash Das

thinkingserious commented:
@dibyadas,

Please add the test in the existing file. Thanks!

w- commented on python_http_client/__init__.py:
```diff
@@ -1 +1,2 @@
 from .client import Client
+from .exceptions import *
```
even though there are many classes., I would change this to an explicit import

thinkingserious replied:
Agreed.

w- commented on python_http_client/exceptions.py:
```diff
@@ -0,0 +1,53 @@
+				503 : ServiceUnavailableError
+}
+
+def handle_error(error):
```
this should probably get_HTTPError(err) and it should return an instance of the specific error class instead of raising it.

dibyadas replied:
Umm.. Can you give an example? I can't visualize this.

w- replied:
this is what i think the function should be

```python
def get_HTTPError(error):
	try:
		http_err_obj = err_dict[error.code](error)
	except KeyError:
		http_err_obj = HttpError(error)		
	return http_err_obj
```

w- commented on python_http_client/exceptions.py:
```diff
@@ -0,0 +1,53 @@
+}
+
+def handle_error(error):
+	exc = err_dict[error.code](error)
```
we want a try/catch here in event that the error.code doesn't match anything in our dict. it should then set err as a generic HTTPError

dibyadas replied:
Actually I just put in a few errors, the ones that SendGrid servers returns, just for showing the approach. If approved, I will add the rest of the error.

thinkingserious replied:
@dibyadas,

Thanks, that would be awesome. I also agree that we should make sure to try/catch for errors we don't expect as @w- illustrated.

w- commented on python_http_client/exceptions.py:
```diff
@@ -0,0 +1,53 @@
+
+def handle_error(error):
+	exc = err_dict[error.code](error)
+	exc.__cause__ = None
```
as per https://www.python.org/dev/peps/pep-3134/
we probably want __cause__ to be the original error.

dibyadas replied:
If we won't to do this then an ugly traceback is produced in Python 3. In this PR, we are raising our custom exception when an urllib.HTTPError is raised. Here, in python 3 , such a thing will be displayed the output.
```
Traceback (most recent call last):
  File "/home/dibya/workspace/python-http-client/python_http_client/client.py", line 141, in _make_request
    return opener.open(request)
  File "/usr/lib/python3.5/urllib/request.py", line 472, in open
    response = meth(req, response)
  File "/usr/lib/python3.5/urllib/request.py", line 582, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/lib/python3.5/urllib/request.py", line 504, in error
    result = self._call_chain(*args)
  File "/usr/lib/python3.5/urllib/request.py", line 444, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.5/urllib/request.py", line 696, in http_error_302
    return self.parent.open(new, timeout=req.timeout)
  File "/usr/lib/python3.5/urllib/request.py", line 472, in open
    response = meth(req, response)
  File "/usr/lib/python3.5/urllib/request.py", line 582, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/lib/python3.5/urllib/request.py", line 510, in error
    return self._call_chain(*args)
  File "/usr/lib/python3.5/urllib/request.py", line 444, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.5/urllib/request.py", line 590, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 401: UNAUTHORIZED

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test.py", line 7, in <module>
    r = a.name.api.get()
  File "/home/dibya/workspace/python-http-client/python_http_client/client.py", line 209, in http_request
    return Response(self._make_request(opener, request))
  File "/home/dibya/workspace/python-http-client/python_http_client/client.py", line 143, in _make_request
    handle_error(err)
  File "/home/dibya/workspace/python-http-client/python_http_client/exceptions.py", line 53, in handle_error
    raise exc
python_http_client.exceptions.UnauthorizedError: HTTP Error 401: UNAUTHORIZED

```
If we set the `__cause__` to None, then only the part below `During handling of the above exception, another exception occurred:` is shown, which is all that should be displayed.

w- replied:
I see. ok. I guess it is preferable to be None

w- commented on python_http_client/exceptions.py:
```diff
@@ -0,0 +1,53 @@
+	def __init__(self,error):
+		self.code = error.code
+		self.reason = error.reason
+		#self.headers = error.headers
```
both headers and body are important to record in the new HTTPError object.

dibyadas replied:
The HTTPError thrown by `urllib` module doesn't have a body attribute and in python versions <= 3.3, they don't have the headers attribute as well. I just commented this out so that this can be discussed further.

w- replied:
it does. you have to read docs and look at the source.

HTTPError is based on URLError which has the following attributes

```python
err.url
err.code
err.hdrs
# to get the body
err.read()

```

dibyadas replied:
Ah! Since it was not mentioned in the documentation, I missed it. Thanks for the tip! :)

w- commented on python_http_client/exceptions.py:
```diff
@@ -0,0 +1,53 @@
+class HTTPError(Exception):
+	''' Base of all other errors'''
+	def __init__(self,error):
+		self.code = error.code
```
keeping in line with other http libraries, I think that self.code should actually be self.status_code

dibyadas replied:
Sure we can. It'll be more evident.

dibyadas commented on python_http_client/client.py:
```diff
@@ -135,7 +137,10 @@ def _make_request(self, opener, request):
+        try:
+            return opener.open(request)
+        except HTTPError as err:
+            handle_error(err)
```
@w- Here, I am handling the urllib.HTTPError.  What you are saying is, we do this:-
```
try:
    return opener.open(request)
except HTTPError as err:
    raise get_HTTPError(err)
```
Am I right?

w- replied:
correct.

thinkingserious commented:
Nice work @dibyadas and @w-! Looks like we are almost there :)

dibyadas commented:
Hi @thinkingserious and @w-  ,
I am not that familiar with adding tests, but I have tried to add some. If have also tried addressing your review comments. Please have a look and suggest changes if any.

With Best Regards,
Dibya Prakash Das

dibyadas commented:
Hi @thinkingserious,
I thought it would be good, if you could take a look at this. I am kind of stuck here. The code coverage passed and then failed for reasons I do not know. I thought maybe you could help me with this.

With Regards,
Dibya Prakash Das

thinkingserious commented:
This looks good to me @dibyadas, what do you think @w-?

w- commented:
I'm ok with this.

thinkingserious commented:
Hello @dibyadas,<br /><br/>Thanks again for the PR!<br />                                              <br />We want to show our appreciation by sending                                              you some swag. Could you please fill out                                              <a href="https://dx.sendgrid.com/swag">this form</a> so we can send it to you? Thanks!                                             <br /><br />Team SendGrid DX

dibyadas commented:
Hi @thinkingserious,
I have filled up the from. Thanks a lot!
With Regards,
Dibya Prakash Das

iandouglas commented:
Per https://github.com/sendgrid/sendgrid-python/issues/323 if we could get examples of these awesome new exceptions in the README so other devs don't need to go surfing into tests to see examples of exceptions and what can cause them, I think that'd be super helpful.

andriisoldatenko commented on python_http_client/exceptions.py:
```diff
@@ -0,0 +1,60 @@
+	pass
+
+class MethodNotAllowedError(HTTPError):
+	pass
```
could you please use spaces and pep8
